### DISCUSSION TITLE
Custom State Machine Behavior

### DISCUSSION THREAD

**Discussion by johnnychoque** [eclipse-edc/Connector#4279] [Q&A]
Hi all,

As part of the integration process of the EDC connector with another system, we need to modify the standard behavior of the Provider's state machine during the transfer process, as shown in the following figure.

![State machine](https://github.com/eclipse-edc/Connector/assets/17832942/87ab084e-e036-4bb5-8d68-5f183272825c)

After the Provider receives the _Transfer Request Message_, it is necessary to execute a specific task of the system to be integrated with the connector. Once this task is completed, the _Transfer Start Message_ will be executed and the standard state machine flow will continue.

Is it possible to interrupt and modify the standard state machine behavior? If so, could you give me some ideas on how to implement it?

Regards,

Johnny

**Comment by paullatzelsperger** [comment]
Hi @johnnychoque I think you could investigate the following two (actually just the first) ways:

1. Use `PendingGuards`: these give you a way to exclude a certain `TransferProcess` from the StateMachine: the TP would be in `"pending"` state, and the state machine doesn't pick it up until the `"pending"` state is reset. This could come in handy for example when some manual human interaction is necessary. 
Technically, you'd have to create an extension that injects an implementation of the `TransferProcessPendingGuard` into the context, so that the state machine picks it up. In that implementation you can implement whatever logic you require.

3. Implement your own `TransferProcessManager`. I only mention this for the sake of completeness. In reality I would _not_ recommend doing that, because it would mean replacing the entire `control-plane-transfer` module. Drilling a hole through your knee sounds more fun. We have never tested this, nor would we support it.

**Reply by ndr-brt** [reply]
for sake of completion, you could also register a `sync` `EventSubscriber` on the `EventRouter`, there, you can write specific logic that will executed when the transfer transitions into `REQUESTED` (catching the `TransferProcessRequested` event). 
By the "sync" attribute, the code will be executed in the same thread/transaction context.
